### PR TITLE
[Feature] Refactor Fast R-CNN

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
- Refactor Fast R-CNN
- Add Dump Proposal Metric

### CODE REVIEW

File: configs/_base_/datasets/coco_detection.py [hunk 1]
```diff
@@ -36,8 +36,9 @@
     dataset=dict(
         type=dataset_type,
         data_root=data_root,
-        ann_file='annotations/instances_train2017.json',
-        data_prefix=dict(img='train2017/'),
+        # indices=50,
```
Comment (BIGWangYuDong) [inline configs/_base_/datasets/coco_detection.py:39]: update

File: mmdet/datasets/transforms/wrappers.py [hunk 2]
```diff
@@ -170,3 +170,113 @@ def __repr__(self):
             format_string += f'{t.__class__.__name__}, '
         format_string += ')'
         return format_string
+
+
+@TRANSFORMS.register_module()
+class ProposalBroadcaster(BaseTransform):
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/wrappers.py:176]: Need to explain the difference with TransformBroadcaster

File: mmdet/datasets/transforms/wrappers.py [hunk 3]
```diff
@@ -170,3 +170,113 @@ def __repr__(self):
...
+        >>>             dict(type='Resize', scale=(1333, 800),
+        >>>                  keep_ratio=True),
+        >>>             dict(type='RandomFlip', prob=0.5),
+        >>>         ]),
+        >>>     dict(type='PackDetInputs')]
+    """
+
+    def __init__(
+        self,
+        transforms: List[Union[dict, Callable[[dict], dict]]],
```
Comment (hhaAndroid) [inline mmdet/datasets/transforms/wrappers.py:212]: support None

File: configs/_base_/models/rpn_r50-caffe-c4.py [hunk 4]
```diff
@@ -53,7 +53,7 @@
                 pos_fraction=0.5,
                 neg_pos_ub=-1,
                 add_gt_as_proposals=False),
-            allowed_border=0,
+            allowed_border=-1,
```
[the remark this answers is not in the document]
Comment (BIGWangYuDong) [reply configs/_base_/models/rpn_r50-caffe-c4.py:56]: This is a legacy issue, this change can align the release AR

File: mmdet/datasets/transforms/wrappers.py (hunk 2 above)
Comment (BIGWangYuDong) [reply mmdet/datasets/transforms/wrappers.py:176]: The TransformBroadcaster cannot update keys that have been updated in the TransformBroadcaster.transforms. 
If it wants to have the same effect as ProposalBroadcaster, it needs to set a lot of mapping keys, which is not easy to expansion.  ProposalBroadcaster is a more concise usage that can reduce the writing of config files

File: configs/fast_rcnn/README.md [hunk 5]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
-    train=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_train2017.pkl'),
-    val=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'),
-    test=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'))
-```
-
-Finally, users can start training the Fast R-CNN.
+The region proposals result can be obtained by setting `DumpProposals` pseudo metric. The dumped results is a `dict(file_name: pred_instance)`.
```
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:18]: The region proposals can be obtained

File: configs/fast_rcnn/README.md [hunk 6]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
+
+  The config of get training image region proposals can be set as below:
+
+  ```python
+  # For training set
+  test_dataloader = dict(
+      dataset=dict(
+          ann_file='data/coco/annotations/instances_train2017.json',
+          data_prefix=dict(img='val2017/')))
+  test_evaluator = dict(
```
Comment (RangiLyu) [inline configs/fast_rcnn/README.md:32]:
```suggestion
  train_dataloader = dict(
      dataset=dict(
          ann_file='data/coco/annotations/instances_train2017.json',
          data_prefix=dict(img='val2017/')))
  train_evaluator = dict(
```

File: configs/fast_rcnn/README.md [hunk 7]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
-        proposal_file='proposals/rpn_r50_fpn_1x_train2017.pkl'),
-    val=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'),
-    test=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'))
-```
-
-Finally, users can start training the Fast R-CNN.
+The region proposals result can be obtained by setting `DumpProposals` pseudo metric. The dumped results is a `dict(file_name: pred_instance)`.
+The `pred_instance` is a `InstanceData` with the ranked bboxes and scores calculate from RPN. We provide example of dumping proposals in [RPN config](../rpn/rpn_r50_fpn_1x_coco.py).
```
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:19]: is a -> is an
with the xxx -> containing the sorted boxes and scores predicted by RPN

File: configs/fast_rcnn/README.md [hunk 8]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
+  # For training set
+  test_dataloader = dict(
+      dataset=dict(
+          ann_file='data/coco/annotations/instances_train2017.json',
+          data_prefix=dict(img='val2017/')))
+  test_evaluator = dict(
+      _delete_=True,
+      type='DumpProposals',
+      output_dir='data/coco/proposals/',
+      proposals_file='rpn_r50_fpn_1x_val2017.pkl')
```
Comment (RangiLyu) [inline configs/fast_rcnn/README.md:36]:
```suggestion
      proposals_file='rpn_r50_fpn_1x_train2017.pkl')
```

File: configs/fast_rcnn/README.md [hunk 9]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
+  ```
+
+  The config of get validation image region proposals can be set as below:
+
+  ```python
+  # For validation set
+  test_evaluator = dict(
```
Comment (RangiLyu) [inline configs/fast_rcnn/README.md:43]:
```suggestion
  val_evaluator = dict(
```

File: configs/fast_rcnn/README.md [hunk 10]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'),
-    test=dict(
-        proposal_file='proposals/rpn_r50_fpn_1x_val2017.pkl'))
-```
-
-Finally, users can start training the Fast R-CNN.
+The region proposals result can be obtained by setting `DumpProposals` pseudo metric. The dumped results is a `dict(file_name: pred_instance)`.
+The `pred_instance` is a `InstanceData` with the ranked bboxes and scores calculate from RPN. We provide example of dumping proposals in [RPN config](../rpn/rpn_r50_fpn_1x_coco.py).
+
+- Firstly, it should be obtained the region proposals in both training and validation (or testing) set.
```
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:21]: Firstly -> First

File: configs/fast_rcnn/README.md [hunk 11]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
+  The config of get validation image region proposals can be set as below:
+
+  ```python
+  # For validation set
+  test_evaluator = dict(
+    _delete_=True,
+    type='DumpProposals',
+    output_dir='data/coco/proposals/',
+    proposals_file='rpn_r50_fpn_1x_val2017.pkl')
+  ```
```
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:48]: Suggest also provide running command here

File: configs/fast_rcnn/README.md [hunk 12]
```diff
@@ -15,49 +15,85 @@ This paper proposes a Fast Region-based Convolutional Network method (Fast R-CNN
...
+
+  Users can follow [test tutorial](https://mmdetection.readthedocs.io/en/3.x/user_guides/test.html) to obtain the region proposals file.
```
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:50]: Users can refer to xxx for more details

File: configs/fast_rcnn/README.md (hunk 5 above)
Comment (ZwwWayne) [inline configs/fast_rcnn/README.md:18]: We could release a region proposal file in our server for users to download.

File: configs/fast_rcnn/README.md (hunk 9 above)
Comment (BIGWangYuDong) [reply configs/fast_rcnn/README.md:43]: should be test_evaluator, so that users can using tools/test.py to dump proposals.

File: configs/fast_rcnn/README.md (hunk 6 above)
Comment (BIGWangYuDong) [reply configs/fast_rcnn/README.md:32]: should be test_evaluator, so that users can using tools/test.py to dump proposals.

File: configs/fast_rcnn/README.md (hunk 5 above)
Comment (BIGWangYuDong) [reply configs/fast_rcnn/README.md:18]: We can upload later

File: mmdet/datasets/transforms/wrappers.py [hunk 13]
```diff
@@ -170,3 +170,111 @@ def __repr__(self):
...
+        cp_data = copy.deepcopy(data)
+        cp_data['gt_bboxes'] = cp_data['proposals']
+        scatters = [data, cp_data]
+        return scatters
+
+    def _apply_transforms(self, inputs: list) -> list:
+        """Apply ``self.transforms``.
+
+        Args:
+            inputs (list[dict, dict]): list of input data.
```
Comment (RangiLyu) [inline mmdet/datasets/transforms/wrappers.py:255]:
```suggestion
            inputs (List[dict]): list of input data.
```

Comment (RangiLyu) [reply mmdet/datasets/transforms/wrappers.py:255]: If the input length is fixed, add an assertion.

**Review by RangiLyu — DISMISSED** [review]
can be merged after updating the docstring

**Review by RangiLyu — APPROVED** [review]
